README.md
# Usefull Extensions 🔌

[![ci][ci_badge]][ci_link] [![pub package][pub_badge]][pub_link] [![style: very good analysis][very_good_analysis_badge]][very_good_analysis_link] [![License: MIT][license_badge]][license_link] ![coverage][coverage_badge]

Generated by the [Very Good CLI](https://github.com/VeryGoodOpenSource/very_good_cli) 🤖

A set of useful extensions and widgets for Flutter and Dart that I use regularly and that I have been collecting through the projects that I have been doing, StackOverflow questions, posts, etc...

## Installing

Add usefull extensions package to your pubspec.yaml file:

```yaml
dependencies:
  usefull_extensions:
```

Import the library into your `.dart` file:

```dart
import 'package:usefull_extensions/usefull_extensions.dart';
```

---

## Extensions

`usefull_extensions` offers different extensions, for Objects, Widgets, Strings, etc...

#### Joined Widgets

This extension allows us to add a separator between each element of a List<Widget>
Usage: `joinWithSeparator`

```dart
[
 const Text('CARLOS'),
 const Text('MOBILE SOLUTIONS'),
].joinWithSeparator(); // You can also specify your own separator Widget
//.joinWithSeparator(const SizedBox(height: 10));
```

#### Spaced Widgets

This extension allows us to add a padding to a whole List<Widget>, by default excludes flex Widgets, such as `Expanded`, `Spacer` or `Flexible`
Usage: `spaced`

```dart
[
  const Text('CARLOS'),
  const Text('MOBILE SOLUTIONS'),
].spaced(); // You can also specify your own EdgeInsetsGeometry
//.spaced(padding: const EdgeInsets.all(10));
```

### Log on Object

This extensions emits a log event of the current object, and returns the length of the output log
Usage: `log`

```dart
final testString = 'My String';
testString.log();
// Outputs -> 'My String' as a log event
```

### Unwrapped String on Object?

This extension allows us to safely unwrap a string, returning `-` by default if it's null
Usage: `unwrappedString`

```dart
null.unwrappedString(); // '-'
null.unwrappedString('🍰'); // '🍰'
'My test'.unwrappedString('?'); // My test
```

### Detailed Where on Map<K, V>

This set of extensions allows us to find an entry based on key, value, or both
Usage: `where`, `whereKey` and `whereValue`

```dart
<String, int>{'John': 20, 'Mary': 21, 'Peter': 20}

..where((key, value) => key.length > 4 && value >= 20) // {Peter: 22}

..whereKey((key) => key.length < 5) // {John: 20, Mary: 21}

..whereValue((value) => value.isEven); // {John: 20, Peter: 22}
```

### Capitalization on String

This extensions allows us to safely capitalize or title-case a String
Usage: `toCapitalized` and `toTitleCase`

```dart
'carlos g'.toCapitalized(); // Carlos g
'carlos g'.toTitleCase(); // Carlos G
```

## Widgets

`usefull_extensions` offers different Widgets

#### Spacer Widgets

A list of spacers, both vertically and horizontally
Usage: `VerticalSpacer`, `HorizontalSpacer` and `VerticalSpacerWithText`

```dart
[
	// Vertical
    VerticalSpacer.xSmall(),
    VerticalSpacer.small(),
    VerticalSpacer.normal(),
    VerticalSpacer.semi(),
    VerticalSpacer.mediumSmall(),
    VerticalSpacer.medium(),
    VerticalSpacer.medium20(),
    VerticalSpacer.mediumLarge(),
    VerticalSpacer.large(),
    VerticalSpacer.xLarge(),
    VerticalSpacer.xxLarge(),
    VerticalSpacer.xxxLarge(),
    
	// Horizontal
    HorizontalSpacer.small(),
    HorizontalSpacer.normal(),
    HorizontalSpacer.semi(),
    HorizontalSpacer.mediumSmall(),
    HorizontalSpacer.medium(),
    HorizontalSpacer.medium20(),
    HorizontalSpacer.mediumLarge(),
    HorizontalSpacer.large(),
    HorizontalSpacer.xLarge(),
    HorizontalSpacer.xxLarge(),
    
	// Vertical with text
    VerticalSpacerWithText(text: 'example text'),
]
```

### ThemeData on BuildContext

This extensions allows us to access a themedata based on the current context

```dart
final theme = context.theme; // Current ThemeData
```


[pub_badge]: https://img.shields.io/pub/v/usefull_extensions.svg?label=usefull_extensions
[pub_link]: https://pub.dev/packages/usefull_extensions

[very_good_analysis_badge]: https://img.shields.io/badge/style-very_good_analysis-B22C89.svg
[very_good_analysis_link]: https://pub.dev/packages/very_good_analysis

[license_badge]: https://img.shields.io/badge/license-BSD%203-green.svg
[license_link]: https://opensource.org/licenses/BSD-3-Clause

[coverage_badge]: https://raw.githubusercontent.com/cgutierr-zgz/usefull_extensions/master/coverage_badge.svg

[ci_badge]: https://github.com/cgutierr-zgz/usefull_extensions/workflows/publish/badge.svg
[ci_link]: https://github.com/cgutierr-zgz/usefull_extensions/actions


CHANGELOG.md
# [1.0.3](https://github.com/cgutierr-zgz/usefull_extensions/commit/e298b547d8aceb7478c9373b34e11f26e8dfee61) (13-07-2022)

### Features

- Added theme of context extension. Usage: `context.theme`

<details >
<summary>Previous versions</summary>

## [1.0.2](https://github.com/cgutierr-zgz/usefull_extensions/commit/8da5654285c24b9923ade4443a00981d5a806a89) (09-07-2022)

### FIX

- Updated README.md broken image on pub.dev ([#7](https://github.com/cgutierr-zgz/usefull_extensions/issues/7)) ([0bcdf4b](https://github.com/cgutierr-zgz/usefull_extensions/commit/0bcdf4b7b2df9311f5f7e949342498558a1be115))

## [1.0.1](https://github.com/cgutierr-zgz/usefull_extensions/commit/c18ce96eb3a868a50edfc17484142fcb2fecc771) (08-07-2022)

### FIX

- Updated README.md broken image on pub.dev ([#7](ht

## [1.0.0](https://github.com/cgutierr-zgz/usefull_extensions/commit/1c7f5d8fc591ee1ef1220ee78d3473fcc55f31e2) (08-07-2022)

### Features

- Added example project ([#1](https://github.com/cgutierr-zgz/usefull_extensions/issues/1)) ([ad6ce5f](https://github.com/cgutierr-zgz/usefull_extensions/commit/ad6ce5f91a60287dd89e068eb8ebfeda228187c7))
- Updated README.md with examples ([#2](https://github.com/cgutierr-zgz/usefull_extensions/issues/2)) ([b007bd4](https://github.com/cgutierr-zgz/usefull_extensions/commit/b007bd4c010c3f81f0aba1d88b1127f2b8c4d907))
- Added 100% test coverage ([#3](https://github.com/cgutierr-zgz/usefull_extensions/issues/3)) ([67ce874](https://github.com/cgutierr-zgz/usefull_extensions/commit/67ce8741cc20042887bfb2345b1b1ddf20cf87a5))
- Added CI/CD pipelines ([#4](https://github.com/cgutierr-zgz/usefull_extensions/issues/4)) ([a0accdc](https://github.com/cgutierr-zgz/usefull_extensions/commit/a0accdcd4bc1764b386ac61c803a272f7bcc6ffc))


## [0.0.1](https://github.com/cgutierr-zgz/usefull_extensions/commit/33844136a4d66310ed0bf8d295bb7f9979f4cfbd) (08-07-2022)

### Features

- Initial release of the `usefull_extensions` package

</details>

test/usefull_extensions_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:usefull_extensions/usefull_extensions.dart';

import 'helpers/pump_app.dart';
import 'helpers/test_helpers.dart';

void main() {
  group('Strings extensions', () {
    const text = 'exception test';
    test('Capitalized text', () {
      final result = text.toCapitalized();

      expect(
        result,
        equals('Exception test'),
      );
    });

    test('Capitalized without text', () {
      const String? test = null;
      final result = test?.toCapitalized();

      expect(
        result,
        equals(null),
      );
    });

    test('TitleCase with text', () {
      final result = text.toTitleCase();

      expect(
        result,
        equals('Exception Test'),
      );
    });

    test('TitleCase without text', () {
      const String? test = null;
      final result = test?.toTitleCase();

      expect(
        result,
        equals(null),
      );
    });
  });

  group('Widgets extensions', () {
    const items = <Widget>[
      Text('a'),
      Text('b'),
      Text('c'),
      Text('d'),
    ];
    test('Adding separator', () {
      final result = items.joinWithSeparator(HorizontalSpacer.semi());

      expect(
        result.length,
        equals(7),
      );
    });

    test('Adding separator on a single widget', () {
      final result = [const Text('a')].joinWithSeparator(
        HorizontalSpacer.semi(),
      );

      expect(
        result.length,
        equals(1),
      );
    });

    test('Adding space should equal same ammount of Widgets, but with padding',
        () {
      final result = items.spaced();

      expect(
        result.length,
        equals(items.length),
      );
    });

    test('Adding a Spacer widget', () {
      final result = [...items, const Spacer()].spaced();

      expect(
        result.length,
        equals(items.length + 1),
      );
    });

    testWidgets(
      'Finds padding when spacing the widgets',
      (tester) async {
        await tester.pumpApp(
          Column(children: items.spaced()),
        );
        expect(find.byType(Padding), findsNWidgets(items.length));
      },
    );
    testWidgets(
      '''Finds paddings when spacing the widgets, doesnt add padding to this to Spacer''',
      (tester) async {
        await tester.pumpApp(
          Column(children: [...items, const Spacer()].spaced()),
        );
        expect(find.byType(Padding), findsNWidgets(items.length));
      },
    );
  });

  group('Object extensions', () {
    test('Log on object', () {
      final strLen = 'MyOutput'.log();

      expect(strLen, 8);
    });
    test('Unwrapped string on object?', () {
      dynamic model;

      model = null.unwrappedString();
      expect(model, '-');

      model = null.unwrappedString('?');
      expect(model, '?');

      model = 'test'.unwrappedString();
      expect(model, 'test');

      model = true;
      expect(model, true);
    });
  });

  group('Object Map<K,V>', () {
    test('where extension', () {
      final people = <String, int>{
        'John': 20,
        'Mary': 21,
        'Peter': 22,
      };

      final subMap = people.where((key, value) => key.length > 4 && value > 20);

      expect(subMap, {'Peter': 22});
    });
    test('whereKey extension', () {
      final people = <String, int>{
        'John': 20,
        'Mary': 21,
        'Peter': 22,
      };

      final subMap = people.whereKey((key) => key.length < 5);

      expect(subMap, {'John': 20, 'Mary': 21});
    });
    test('whereValue extension', () {
      final people = <String, int>{
        'John': 20,
        'Mary': 21,
        'Peter': 22,
      };

      final subMap = people.whereValue((value) => value.isEven);

      expect(subMap, {'John': 20, 'Peter': 22});
    });
  });
  group('Paddings', () {
    testWidgets(
      'Vetical Spacer',
      (tester) async {
        await tester.pumpApp(
          Column(
            children: [
              const Text('1'),
              VerticalSpacer.xSmall(),
              VerticalSpacer.small(),
              VerticalSpacer.normal(),
              VerticalSpacer.semi(),
              VerticalSpacer.mediumSmall(),
              VerticalSpacer.medium(),
              VerticalSpacer.medium20(),
              VerticalSpacer.mediumLarge(),
              VerticalSpacer.large(),
              VerticalSpacer.xLarge(),
              VerticalSpacer.xxLarge(),
              VerticalSpacer.xxxLarge(),
              // 12 types of spacers
              const Text('2')
            ],
          ),
        );
        expect(find.byType(VerticalSpacer), findsNWidgets(12));
      },
    );
    testWidgets(
      'Horizonal Spacer',
      (tester) async {
        await tester.pumpApp(
          Column(
            children: [
              const Text('1'),
              HorizontalSpacer.xSmall(),
              HorizontalSpacer.small(),
              HorizontalSpacer.normal(),
              HorizontalSpacer.semi(),
              HorizontalSpacer.mediumSmall(),
              HorizontalSpacer.medium(),
              HorizontalSpacer.medium20(),
              HorizontalSpacer.mediumLarge(),
              HorizontalSpacer.large(),
              HorizontalSpacer.xLarge(),
              HorizontalSpacer.xxLarge(),
              HorizontalSpacer.xxxLarge(),
              // 12 types of spacers
              const Text('2')
            ],
          ),
        );
        expect(find.byType(HorizontalSpacer), findsNWidgets(12));
      },
    );
    testWidgets(
      'Horizonal Spacer with text',
      (tester) async {
        FlutterError.onError = ignoreOverflowErrors;
        const hText = 'Horizontal spacer text!';
        await tester.pumpApp(
          const VerticalSpacerWithText(
            text: hText,
            color: Colors.red,
          ),
        );
        expect(find.byType(VerticalSpacerWithText), findsNWidgets(1));
        expect(find.text(hText), findsOneWidget);
      },
    );
  });
  group('Context extensions', () {
    testWidgets(
      'theme of context',
      (tester) async {
        // FlutterError.onError = ignoreOverflowErrors;
        final theme = ThemeData(brightness: Brightness.light);
        await tester.pumpWidget(
          MaterialApp(
            theme: ThemeData(brightness: Brightness.dark),
            home: const Scaffold(),
          ),
        );

        // Capture a BuildContext object
        final BuildContext context = tester.element(find.byType(MaterialApp));

        expect(context.theme.brightness, theme.brightness);
      },
    );
  });
}


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:usefull_extensions/usefull_extensions.dart';

void main() {
  // EXTENSIONS
  // log
  'My String'.log(); // Outputs 'My String' as log

  // map
  <String, int>{'John': 20, 'Mary': 21, 'Peter': 20}
    // {Peter: 22}
    ..where((key, value) => key.length > 4 && value >= 20).log()
    // {John: 20, Mary: 21}
    ..whereKey((key) => key.length < 5).log()
    // {John: 20, Peter: 22}
    ..whereValue((value) => value.isEven).log();

  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Usefull extensions and Widgets',
      home: Scaffold(
        backgroundColor: Colors.grey[200],
        body: Center(
          child: _SeparatedColumn(),
        ),
      ),
    );
  }
}

class _SeparatedColumn extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      child: Column(
        children: [
          _SeparatedRow(),
          const VerticalSpacerWithText(text: 'V-Spacer with text'),
          _SeparatedRow(),
          /* Converts to - */
          VerticalSpacerWithText(text: null.unwrappedString()),
          _SeparatedRow(),
          /* Converts to ? */
          VerticalSpacerWithText(text: null.unwrappedString('?')),
          _SeparatedRow(),
          /* Converts to Carlos g */
          VerticalSpacerWithText(text: 'carlos g'.toCapitalized()),
          _SeparatedRow(),
          /* Converts to Carlos G */
          VerticalSpacerWithText(text: 'carlos g'.toTitleCase()),
          _SeparatedRow(),
        ]
            /* Separates Vertically */
            .joinWithSeparator(VerticalSpacer.large()),
      ),
    );
  }
}

class _SeparatedRow extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      scrollDirection: Axis.horizontal,
      child: Row(
        children: [
          _RandomContainer(),
          _RandomContainer(),
          _RandomContainer(),
          _RandomContainer(),
        ]
            /* Separates Horizontally */
            .joinWithSeparator(HorizontalSpacer.large())
            /* Adds Paddings */
            .spaced(),
      ),
    );
  }
}

class _RandomContainer extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final rand = Random();
    final color = Colors.primaries[rand.nextInt(Colors.primaries.length)];
    final size = (rand.nextInt(100) + 50).toDouble();

    return Container(color: color, height: size, width: size);
  }
}


lib/usefull_extensions.dart
import 'dart:developer' as devtools;
import 'package:flutter/material.dart';

/// A collection of usefull extensions on [String]
extension StringCasingExtension on String {
  /// Converts the first letter of a [String] into uppercase
  /// or returns '' if null
  ///
  /// Example:
  /// ```dart
  /// 'carlos gutiérrez'.toCapitalized(); // Carlos gutiérrez
  /// ```
  String toCapitalized() =>
      length > 0 ? '${this[0].toUpperCase()}${substring(1).toLowerCase()}' : '';

  /// Converts all the first letters of each words of a given [String] into
  /// uppercase
  ///
  /// Example:
  /// ```dart
  /// 'carlos gutiérrez'.toTitleCase(); // Carlos Gutiérrez
  /// ```
  /// See also:
  ///
  ///  * [toCapitalized]
  String toTitleCase() => replaceAll(RegExp(' +'), ' ')
      .split(' ')
      .map((str) => str.toCapitalized())
      .join(' ');
}

/// A collection of usefull extensions on [List<Widget>]
extension JoinedWidgets on List<Widget> {
  /// Adds a specific type of [Widget] in between a list of Widgets
  /// This can be usefull to add some height in between Widgets
  /// without the need of writing it multiple times
  ///
  /// Example:
  /// ```dart
  /// [
  ///  const Text('CARLOS'),
  ///  const Text('MOBILE SOLUTIONS'),
  /// ].joinWithSeparator(const SizedBox(height: 10));
  /// ```
  List<Widget> joinWithSeparator(Widget? separator) {
    return length > 1
        ? (take(length - 1)
            .map(
              (widget) =>
                  [widget, separator ?? const SizedBox(height: 24, width: 24)],
            )
            .expand((widget) => widget)
            .toList()
          ..add(last))
        : this;
  }

  /// Adds the same ammount of padding to a list of Widgets
  /// This can be usefull if you need to have some of the widgets of a list
  /// spaced, and some don't
  ///
  /// Example:
  /// ```dart
  /// [
  ///   const Text('CARLOS'),
  ///   const Text('MOBILE SOLUTIONS'),
  /// ].spaced();
  /// // or
  /// ].spaced(padding: const EdgeInsets.all(10));
  /// ```
  List<Widget> spaced({EdgeInsetsGeometry? padding, bool excludeFlex = true}) {
    final spacedWidgets = <Widget>[];
    for (final w in this) {
      if (excludeFlex && (w is Expanded || w is Spacer || w is Flexible)) {
        spacedWidgets.add(w);
      } else {
        spacedWidgets.add(
          Padding(
            padding: padding ?? const EdgeInsets.fromLTRB(24, 0, 24, 0),
            child: w,
          ),
        );
      }
    }
    return spacedWidgets;
  }
}

/// {@template log_extension}
/// Emit a log event of the current object
/// {@endtemplate}
extension Log on Object {
  /// {@macro log_extension}
  /// Example:
  /// ```dart
  /// final model = Model();
  /// model.log(); // outputs the model toString in a log event
  ///              //and returns the string length
  /// ```
  int log() {
    final str = toString();
    devtools.log(str);

    // returns the length of the ouput
    return str.length;
  }
}

/// {@template unwrapped_extension}
/// Returns the current object.toString() in a safe way
/// {@endtemplate}
extension UnwrappedString on Object? {
  /// {@macro unwrapped_extension}
  /// Example:
  /// ```dart
  /// final Model? model = // My call which returns Model?
  /// // Returns the actual toString if its not null
  /// // or
  /// model.unwrappedString() // '-'
  /// model.unwrappedString('?') // '?'
  /// ```
  String unwrappedString([String replacement = '-']) => this == null
      ? replacement
      : (this as String?)!.isEmpty
          ? replacement
          : toString();
}

/// {@template map_extension}
/// Allows us to find an entry
/// {@endtemplate}
extension DetailedWhere<K, V> on Map<K, V> {
  /// {@macro map_extension}
  /// based on key and value
  ///
  /// Example:
  /// ```dart
  /// people.where((key, value) => key.length > 4 && value >= 20);
  /// // {Peter: 22}
  ///
  /// const Map<String, int> people = {'John': 20, 'Mary': 21, 'Peter':20};
  /// ```
  Map<K, V> where(bool Function(K key, V value) f) => Map<K, V>.fromEntries(
        entries.where((entry) => f(entry.key, entry.value)),
      );

  /// {@macro map_extension}
  /// based on key
  ///
  /// Example:
  /// ```dart
  /// people.whereKey((key) => key.length < 5);
  /// // {John: 20, Mary: 21}
  ///
  /// const Map<String, int> people = {'John': 20, 'Mary': 21, 'Peter':20};
  /// ```
  Map<K, V> whereKey(bool Function(K key) f) =>
      {...where((key, value) => f(key))};

  /// {@macro map_extension}
  /// based on  value
  ///
  /// Example:
  /// ```dart
  /// people.whereValue((value) => value.isEven);
  /// // {John: 20, Peter: 22}
  ///
  /// const Map<String, int> people = {'John': 20, 'Mary': 21, 'Peter':20};
  /// ```
  Map<K, V> whereValue(bool Function(V value) f) =>
      {...where((key, value) => f(value))};
}

/// {@template dimens}
/// Defines the default Dimens for padding to be used as guidelines
///
/// Example:
/// ```dart
/// SizedBox(height: Dimenx.paddingSemi) // 10
/// ```
/// {@endtemplate}
abstract class Dimens {
  /// padding of [2.0]
  /// {@macro map_extension}
  static const double paddingXSmall = 2;

  /// padding of [4.0]
  /// {@macro map_extension}
  static const double paddingSmall = 4;

  /// padding of [8.0]
  /// {@macro map_extension}
  static const double paddingDefault = 8;

  /// padding of [10.0]
  /// {@macro map_extension}
  static const double paddingSemi = 10;

  /// padding of [12.0]
  /// {@macro map_extension}
  static const double paddingMediumSmall = 12;

  /// padding of [16.0]
  /// {@macro map_extension}
  static const double paddingMedium = 16;

  /// padding of [20.0]
  /// {@macro map_extension}
  static const double paddingMedium20 = 20;

  /// padding of [24.0]
  /// {@macro map_extension}
  static const double paddingMediumLarge = 24;

  /// padding of [32.0]
  /// {@macro map_extension}
  static const double paddingLarge = 32;

  /// padding of [36.0]
  /// {@macro map_extension}
  static const double paddingXLarge = 36;

  /// padding of [48.0]
  /// {@macro map_extension}
  static const double paddingXXLarge = 48;

  /// padding of [64.0]
  /// {@macro map_extension}
  static const double paddingXXXLarge = 64;
}

/// [VerticalSpacer] creates a vertical separation between Widgets
/// the same way you would use a SizedBox, but with predefined values
/// from [Dimens]
///
/// Example:
/// ```dart
///    VerticalSpacer.xSmall(),
///    VerticalSpacer.small(),
///    VerticalSpacer.normal(),
///    VerticalSpacer.semi(),
///    VerticalSpacer.mediumSmall(),
///    VerticalSpacer.medium(),
///    VerticalSpacer.medium20(),
///    VerticalSpacer.mediumLarge(),
///    VerticalSpacer.large(),
///    VerticalSpacer.xLarge(),
///    VerticalSpacer.xxLarge(),
///    VerticalSpacer.xxxLarge(),
/// ```
///
/// See also:
///
///  * [Dimens]
///  * [HorizontalSpacer]
///  * [VerticalSpacerWithText]
class VerticalSpacer extends SizedBox {
  /// [VerticalSpacer] with custom paddings
  const VerticalSpacer({super.key, required double super.height});

  /// [VerticalSpacer] with [2.0] as padding
  factory VerticalSpacer.xSmall() =>
      const VerticalSpacer(height: Dimens.paddingXSmall);

  /// [VerticalSpacer] with [4.0] as padding
  factory VerticalSpacer.small() =>
      const VerticalSpacer(height: Dimens.paddingSmall);

  /// [VerticalSpacer] with [8.0] as padding
  factory VerticalSpacer.normal() =>
      const VerticalSpacer(height: Dimens.paddingDefault);

  /// [VerticalSpacer] with [10.0] as padding
  factory VerticalSpacer.semi() =>
      const VerticalSpacer(height: Dimens.paddingSemi);

  /// [VerticalSpacer] with [12.0] as padding
  factory VerticalSpacer.mediumSmall() =>
      const VerticalSpacer(height: Dimens.paddingMediumSmall);

  /// [VerticalSpacer] with [16.0] as padding
  factory VerticalSpacer.medium() =>
      const VerticalSpacer(height: Dimens.paddingMedium);

  /// [VerticalSpacer] with [20.0] as padding
  factory VerticalSpacer.medium20() =>
      const VerticalSpacer(height: Dimens.paddingMedium20);

  /// [VerticalSpacer] with [24.0] as padding
  factory VerticalSpacer.mediumLarge() =>
      const VerticalSpacer(height: Dimens.paddingMediumLarge);

  /// [VerticalSpacer] with [32.0] as padding
  factory VerticalSpacer.large() =>
      const VerticalSpacer(height: Dimens.paddingLarge);

  /// [VerticalSpacer] with [36.0] as padding
  factory VerticalSpacer.xLarge() =>
      const VerticalSpacer(height: Dimens.paddingXLarge);

  /// [VerticalSpacer] with [48.0] as padding
  factory VerticalSpacer.xxLarge() =>
      const VerticalSpacer(height: Dimens.paddingXXLarge);

  /// [VerticalSpacer] with [64.0] as padding
  factory VerticalSpacer.xxxLarge() =>
      const VerticalSpacer(height: Dimens.paddingXXXLarge);
}

/// [HorizontalSpacer] creates a horizontal separation between Widgets
/// the same way you would use a SizedBox, but with predefined values
/// from [Dimens]
///
/// Example:
/// ```dart
///    HorizontalSpacer.small(),
///    HorizontalSpacer.normal(),
///    HorizontalSpacer.semi(),
///    HorizontalSpacer.mediumSmall(),
///    HorizontalSpacer.medium(),
///    HorizontalSpacer.medium20(),
///    HorizontalSpacer.mediumLarge(),
///    HorizontalSpacer.large(),
///    HorizontalSpacer.xLarge(),
///    HorizontalSpacer.xxLarge(),
/// ```
///
/// See also:
///
///  * [Dimens]
///  * [VerticalSpacer]
///  * [VerticalSpacerWithText]
class HorizontalSpacer extends SizedBox {
  /// [HorizontalSpacer] with custom paddings
  const HorizontalSpacer({super.key, required double super.width});

  /// [HorizontalSpacer] with [2.0] as padding
  factory HorizontalSpacer.xSmall() =>
      const HorizontalSpacer(width: Dimens.paddingXSmall);

  /// [HorizontalSpacer] with [4.0] as padding
  factory HorizontalSpacer.small() =>
      const HorizontalSpacer(width: Dimens.paddingSmall);

  /// [HorizontalSpacer] with [8.0] as padding
  factory HorizontalSpacer.normal() =>
      const HorizontalSpacer(width: Dimens.paddingDefault);

  /// [HorizontalSpacer] with [10.0] as padding
  factory HorizontalSpacer.semi() =>
      const HorizontalSpacer(width: Dimens.paddingSemi);

  /// [HorizontalSpacer] with [12.0] as padding
  factory HorizontalSpacer.mediumSmall() =>
      const HorizontalSpacer(width: Dimens.paddingMediumSmall);

  /// [HorizontalSpacer] with [16.0] as padding
  factory HorizontalSpacer.medium() =>
      const HorizontalSpacer(width: Dimens.paddingMedium);

  /// [HorizontalSpacer] with [20.0] as padding
  factory HorizontalSpacer.medium20() =>
      const HorizontalSpacer(width: Dimens.paddingMedium20);

  /// [HorizontalSpacer] with [24.0] as padding
  factory HorizontalSpacer.mediumLarge() =>
      const HorizontalSpacer(width: Dimens.paddingMediumLarge);

  /// [HorizontalSpacer] with [32.0] as padding
  factory HorizontalSpacer.large() =>
      const HorizontalSpacer(width: Dimens.paddingLarge);

  /// [HorizontalSpacer] with [36.0] as padding
  factory HorizontalSpacer.xLarge() =>
      const HorizontalSpacer(width: Dimens.paddingXLarge);

  /// [HorizontalSpacer] with [48.0] as padding
  factory HorizontalSpacer.xxLarge() =>
      const HorizontalSpacer(width: Dimens.paddingXXLarge);

  /// [HorizontalSpacer] with [64.0] as padding
  factory HorizontalSpacer.xxxLarge() =>
      const HorizontalSpacer(width: Dimens.paddingXXXLarge);
}

/// [VerticalSpacerWithText] creates a horizontal separation between Widgets
/// with a text in between
///
/// Example:
/// ```dart
///    VerticalSpacerWithText(
///      text: 'example text',
///      color: Colors.green,
///    );
/// ```
///
/// See also:
///
///  * [Dimens]
///  * [VerticalSpacer]
///  * [VerticalSpacerWithText]
class VerticalSpacerWithText extends StatelessWidget {
  /// [HorizontalSpacer] with custom attributes
  const VerticalSpacerWithText({
    super.key,
    required this.text,
    this.textStyle,
    this.color,
    this.height = 36,
    this.thickness = 1.0,
  });

  /// Text that will be displayed in between widgets
  final String text;

  /// Style of the text in between dividers
  final TextStyle? textStyle;

  /// Color of the separation in between widgets
  final Color? color;

  /// Height of the divider
  final double? height;

  /// Thickness of the divider
  final double? thickness;

  @override
  Widget build(BuildContext context) {
    return Row(
      children: [
        _line(),
        VerticalSpacer.small(),
        Text(text, style: textStyle),
        VerticalSpacer.small(),
        _line(),
      ],
    );
  }

  Widget _line() => Expanded(
        child: Container(
          margin: const EdgeInsets.all(Dimens.paddingMedium20),
          child: Divider(thickness: thickness, color: color, height: height),
        ),
      );
}

/// {@template theme_extension}
/// Allows access to the current theme data
/// {@endtemplate}
extension ThemeExtension on BuildContext {
  /// {@macro theme_extension}
  /// Example:
  /// ```dart
  /// final theme = context.theme; // Access current theme data
  /// ```
  ThemeData get theme => Theme.of(this);
}
